auth/provider.go

```diff
@@ -56,26 +61,280 @@ func (r *NoAuthScope) Allow(_ string, _ string) (ok bool) {
 
 //
 // NewKeycloak builds a new Keycloak auth provider.
-func NewKeycloak(host, realm, id, secret string) (k Keycloak) {
+func NewKeycloak(host, realm, id, secret, admin, pass, adminRealm string) (k Keycloak) {
 	client := gocloak.NewClient(host)
 	k = Keycloak{
-		host:   host,
-		realm:  realm,
-		id:     id,
-		secret: secret,
-		client: client,
+		host:       host,
+		realm:      realm,
+		id:         id,
+		secret:     secret,
+		client:     client,
+		admin:      admin,
+		pass:       pass,
+		adminRealm: adminRealm,
 	}
 	return
 }
 
 //
 // Keycloak auth provider
 type Keycloak struct {
-	client gocloak.GoCloak
-	host   string
-	realm  string
-	id     string
-	secret string
+	client     gocloak.GoCloak
+	host       string
+	realm      string
+	id         string
+	secret     string
+	admin      string
+	pass       string
+	adminRealm string
+	token      *gocloak.JWT
+}
+
+//
+// Realm is a container for the users,
+// scopes, and roles that exist in the
+// hub's keycloak realm.
+type Realm struct {
+	Users  map[string]gocloak.User
+	Scopes map[string]gocloak.ClientScope
+	Roles  map[string]gocloak.Role
+}
+
+//
+// Reconcile ensures that Keycloak is configured
+// with the scopes, roles, and users required by
+// the hub.
+func (r *Keycloak) Reconcile() (err error) {
+	err = r.login()
+	if err != nil {
+		return
+	}
+
+	realm, err := r.loadRealm()
+	if err != nil {
+		return
+	}
+
+	err = r.ensureRoles(realm)
+	if err != nil {
+		return
+	}
+
+	err = r.ensureUsers(realm)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+//
+// loadRealm loads all the scopes, roles, and users from
+// the hub keycloak realm and populates a Realm struct.
+func (r *Keycloak) loadRealm() (realm *Realm, err error) {
+	realm = &Realm{}
+	realm.Scopes, err = r.scopeMap()
+	if err != nil {
+		return
+	}
+	realm.Roles, err = r.realmRoleMap()
+	if err != nil {
+		return
+	}
+	realm.Users, err = r.userMap()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+//
+// ensureUsers ensures that the hub users exist and have the necessary roles.
+func (r *Keycloak) ensureUsers(realm *Realm) (err error) {
+	users, err := LoadUsers(Settings.Auth.UserPath)
+	if err != nil {
+		return
+	}
+
+	for _, user := range users {
+		if _, ok := realm.Users[user.Name]; !ok {
```

Comment:
This is just a personal nit: but I find the `ok` convention in Go rather odd and un-descriptive and feel that something more descriptive like `found` would be better.